### ISSUE TITLE
(Missing) Index Usage Analytics

### ISSUE THREAD

**Issue by guyellis** [JanusGraph/janusgraph#105]
This warning:

```
75604240 [gremlin-server-exec-4] WARN
com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx  - 
Query requires iterating over all vertices [(~label = Person AND id = 123)].
For better performance, use indexes
```

I believe is generated by [this line](https://github.com/JanusGraph/janusgraph/blob/ec441ac6d87f3359c9740075786ec6de3651745e/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java#L1276).

I am thinking about modifying that section of code to try and collect stats that will aggregate missing index information to provide value/cost data in adding indexes to a graph.

e.g. it would collect the data and then summarize like this:

```
Over X days missed indexes resulted in:

[(~label = Person AND id = 123)] - 3,500 queries
[(~label = Person AND name = 'aaa')] - 320 queries
[(~label = Person AND profession = 'developer')] - 2 queries
```

The idea is to discover which missing indexes would be more used than others and are worth considering for addition.

There is also the consideration in adding analytics logging to the other side of the `if...else` which records indexes being used to discover existing indexes that are under-utilized and might be removed to reduce their maintenance cost.

Questions:

1. Would this sort of thing be useful for others and should I consider a PR if I am successful in conditionally adding this?
2. Where would you store and analyze this information? (Sending external keeps the footprint of Janus Graph small but requires users to add an external "receiving" end-point etc. so more difficult to use. Keeping in memory would require more code in the code base to maintain but could be configured to only run when the user wants the analytics.)
3. Does something like this already exist? Or how do you currently analyze and optimize indexes in your installation?

**Comment by amcp** [comment]
I would be cautious of the performance effect of such a change (because it would get executed in each traversal step). To control the CPU and memory usage side effects, you could limit the output of your new metrics to the top ten missing indexes, and sample call data (instead of record all of it). You could use metrics and create new metrics ad-hoc to track each missing index. Currently, you have the option to [disable queries not supported by indexes](http://docs.janusgraph.org/0.1.0-SNAPSHOT/config-ref.html#_query) using the `query.force-index` configuration option, so the first time you execute a traversal not supported by an index you will find out and have the opportunity to add said index.

**Comment by guyellis** [comment]
@amcp I agree that there might be a performance effect while collecting that. It's something that one would enable for a short period of time (hours or days) to assist in fine tuning and usage discovery.

Thanks for the tip on `force-index`. I've seen that in the docs but not tried it yet.
